promql/testdata/functions.test

```diff
@@ -67,31 +246,103 @@ eval instant at 15m changes(x[15m])
 
 clear
 
-# Tests for increase().
-load 5m
+# Tests for increase()/xincrease()/xincrease0()//xrate()/xrate0().
+load 5s
 	http_requests{path="/foo"}	0+10x10
-	http_requests{path="/bar"}	0+10x5 0+10x5
+	http_requests{path="/bar"}	0+10x5 0+10x4
 
 # Tests for increase().
-eval instant at 50m increase(http_requests[50m])
+eval instant at 50s increase(http_requests[50s])
 	{path="/foo"} 100
 	{path="/bar"}  90
 
-eval instant at 50m increase(http_requests[100m])
+eval instant at 50s increase(http_requests[100s])
 	{path="/foo"} 100
 	{path="/bar"}  90
 
+# Tests for xincrease()/xincrease0().
+eval instant at 50s xincrease(http_requests[50s])
+	{path="/foo"} 100
+	{path="/bar"}  90
+
+eval instant at 50s xincrease0(http_requests[50s])
+	{path="/foo"} 100
+	{path="/bar"}  90
+
+eval instant at 50s xincrease(http_requests[5s])
+	{path="/foo"} 10
+	{path="/bar"} 10
+
+eval instant at 50s xincrease0(http_requests[5s])
+	{path="/foo"} 100
+	{path="/bar"} 40
+
+eval instant at 50s xincrease(http_requests[3s])
+	{path="/foo"} 6
+	{path="/bar"} 6
+
+eval instant at 50s xincrease0(http_requests[3s])
+	{path="/foo"} 60
+	{path="/bar"} 24
+
+eval instant at 49s xincrease(http_requests[3s])
+
+eval instant at 49s xincrease0(http_requests[3s])
+
+# Tests for xrate()/xrate0().
+eval instant at 50s xrate(http_requests[50s])
+	{path="/foo"} 2
+	{path="/bar"} 1.8
+
+eval instant at 50s xrate0(http_requests[50s])
+	{path="/foo"} 2
+	{path="/bar"} 1.8
+
+eval instant at 50s xrate(http_requests[100s])
+	{path="/foo"} 1
+	{path="/bar"} 0.9
+
+eval instant at 50s xrate0(http_requests[100s])
+	{path="/foo"} 2
+	{path="/bar"} 1.8
+
+eval instant at 50s xrate(http_requests[5s])
+	{path="/foo"} 2
+	{path="/bar"} 2
+
+eval instant at 50s xrate0(http_requests[5s])
+	{path="/foo"} 20
+	{path="/bar"} 8
+
+eval instant at 50s xrate(http_requests[3s])
+	{path="/foo"} 2
+	{path="/bar"} 2
+
+eval instant at 50s xrate0(http_requests[3s])
+	{path="/foo"} 20
+	{path="/bar"} 8
+
+eval instant at 49s xrate(http_requests[3s])
+
+eval instant at 49s xrate0(http_requests[3s])
+
 clear
 
-# Test for increase() with counter reset.
+# Test for increase()/xincrease/xincrease0 with counter reset.
 # When the counter is reset, it always starts at 0.
 # So the sequence 3 2 (decreasing counter = reset) is interpreted the same as 3 0 1 2.
 # Prometheus assumes it missed the intermediate values 0 and 1.
 load 5m
 	http_requests{path="/foo"}	0 1 2 3 2 3 4
 
 eval instant at 30m increase(http_requests[30m])
-    {path="/foo"} 7
+	{path="/foo"} 7
+
+eval instant at 30m xincrease(http_requests[30m])
+	{path="/foo"} 7
+
+eval instant at 30m xincrease0(http_requests[30m])
+	{path="/foo"} 5
```

Comment:
Could you do a breakdown why these values are different?

Reply:
Frankly I'm not sure. So far I've just pasted in the test values to make the tests green. That has the limited benefit of proving that that the code hasn't drifted, but it doesn't in any way prove it's correct. For that I'm going to need to understand this whole test DSL.